patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs

```diff
@@ -436,4 +448,40 @@ private void Upgrade()
 		Open();
 		// Read contract fulfilled
 	}
+
+	internal bool VerifyHash()
+	{
+		IDisposable handle = null;
+		if (fileStream == null)
+			handle = Open();
+
+		if (hashStartPos == 0)
+			return false;
+
+		if (hashHasBeenTaken)
+			return hashMatches;
+
+		var oldPos = fileStream.Position;
+		fileStream.Position = hashStartPos;
+		var result = Hash.SequenceEqual(sha1.ComputeHash(fileStream));
+		fileStream.Position = oldPos;
```

Comment:
No need to put `Position` back, but you should assert that `sharedEntryReadStream == null` just like in `GetStream`. Or just open a new `FileStream`, I don't think there's any real benefit to reusing the existing one for this.